src/GraphQLExecutorWithSubscriptions.ts

```diff
@@ -0,0 +1,242 @@
+
+export interface TriggerConfig {
+    channelOptions?: Object;
+    filter?: Function;
+}
+
+export interface TriggerMap {
+    [triggerName: string]: TriggerConfig;
+}
+
+export interface SetupFunction {
+    // TODO: Resolve the options typing
+    (options: any, args: {[key: string]: any}, subscriptionName: string): TriggerMap;
+    //(options: SubscriptionOptions, args: {[key: string]: any}, subscriptionName: string): TriggerMap;
+}
+
+export interface SetupFunctions {
+    [subscriptionName: string]: SetupFunction;
+}
+
+export interface IObservable<T> {
+  subscribe(observer: {
+    next?: (v: T) => void;
+    error?: (e: Error) => void;
+    complete?: () => void
+  }): { unsubscribe: () => void };
+}
+
+export class ValidationError extends Error {
+    errors: Array<GraphQLError>;
+    message: string;
+
+    constructor(errors){
+        super();
+        this.errors = errors;
+        this.message = 'Subscription query has validation errors';
+    }
+}
+
+function ObservableOfErrors(errors: Error[]) {
+  return {
+    subscribe: (observer) => {
+      observer.next && observer.next({ errors });
+      observer.complete && observer.complete();
+      return { unsubscribe: () => {/* noop */} };
+    },
+  };
+}
+
+export class GraphQLExecutorWithSubscriptions {
+  private pubsub: PubSubEngine;
+  private setupFunctions: SetupFunctions
+
+  constructor(options: { pubsub: PubSubEngine, setupFunctions: SetupFunctions }) {
+    this.setupFunctions = options.setupFunctions;
+    this.pubsub = options.pubsub;
+  }
+
+  public executeReactive(
+    schema: GraphQLSchema,
+    document: DocumentNode,
+    rootValue?: any,
+    contextValue?: any,
+    variableValues?: {[key: string]: any},
+    operationName?: string,
+  ) {
+    const errors = validate(schema, document);
+    if ( errors.length > 0 ) {
+      return ObservableOfErrors([new ValidationError(errors)]);
+    }
+
+    const operationAST = getOperationAST(document, operationName);
+    if ( null === operationAST ) {
+      return ObservableOfErrors([new Error(`could not parse operation on query`)]);
+    }
+
+    if ( operationAST.operation === 'subscription' ) {
+      return this.handleSubscription(schema, document, rootValue, contextValue, variableValues, operationName);
+    } else {
+      return this.handleStatic(schema, document, rootValue, contextValue, variableValues, operationName);
+    }
+  }
+
+  public execute(
+    schema: GraphQLSchema,
+    document: DocumentNode,
+    rootValue?: any,
+    contextValue?: any,
+    variableValues?: {[key: string]: any},
+    operationName?: string,
+  ): Promise<ExecutionResult> {
+    return new Promise((resolve, reject) => {
+      const sub = this.executeReactive(schema, document, rootValue, contextValue, variableValues, operationName)
+        .subscribe({
+          next: (v) => {
+            resolve(v)
+            process.nextTick(() => sub.unsubscribe());
+          },
+          error: (e) => reject(e),
```

Comment:
Thats on the observer's duty, it might just unsubscribe or handle the error if it wants too, but we shouldn't unsubscribe for the user